packages/data/src/factory.js

```diff
@@ -39,17 +39,22 @@
  * @return {Function} Registry selector that can be registered with a store.
  */
 export function createRegistrySelector( registrySelector ) {
-	let selector;
-	let lastRegistry;
+	const selectorsByRegistry = new WeakMap();
 	// Create a selector function that is bound to the registry referenced by `selector.registry`
 	// and that has the same API as a regular selector. Binding it in such a way makes it
 	// possible to call the selector directly from another selector.
 	const wrappedSelector = ( ...args ) => {
-		if ( ! selector || lastRegistry !== wrappedSelector.registry ) {
-			selector = registrySelector( wrappedSelector.registry.select );
-			lastRegistry = wrappedSelector.registry;
+		// We want to make sure the cache persists even when new registry
+		// instances are created. For example patterns create their own editors
+		// with their own core/block-editor stores, so we should keep track of
+		// the cache for each registry instance.
+		if ( ! selectorsByRegistry.has( wrappedSelector.registry ) ) {
+			selectorsByRegistry.set(
+				wrappedSelector.registry,
+				registrySelector( wrappedSelector.registry.select )
+			);
 		}
-		return selector( ...args );
+		return selectorsByRegistry.get( wrappedSelector.registry )( ...args );
```

Comment:
Let's write the code in such a way that there is ony one `.get` map lookup, not the two `.has` and `.get`. This will be called all the time, so a micro-opt is justified.

Reply:
Adjusted :)